### DISCUSSION TITLE
generic0-model

### DISCUSSION THREAD

**Discussion by saerbar** [inlabru-org/inlabru#243] [Q&A] [answered]
Hi,

I am trying to implement a generic0-model where the precision matrix comes from an spde model, but I'm not sure how to specify it when not using inla.spde.make.index.

This is how I have implemented the model in inla

```
spde_model <- inla.spde2.matern(spde_mesh, alpha=2)

Q <- inla.spde2.precision(spde_model, theta=Q_params)

A <- inla.spde.make.A(mesh=spde_mesh, loc=coords,
                        group = data$year_idx)
  
spde_indices <- inla.spde.make.index("spde_idx", n.spde = n_spde_mesh, 
                                     n.group=length(unique(data_spde$year_idx)))

stack = inla.stack(
    data = data['y'],
    A = list(A, 1),
    effects = list(spde_indices,
                   list(Intercept = 1,
                        year_idx = data$year_idx,
                        ... other covariates ...)),
    tag = "spde_data"
  )

f <- y ~ f(spde_idx, model = "generic0", Cmatrix = Q, 
                group = spde_idx.group, control.group = list(model = "iid")) + ... other covariates ...

fitted_model <- inla(formula = f, data = inla.stack.data(stack), ... other settings ...)
```

Can I fit this model also in inlabru? What do I do then to replace 'stack' and 'spde_indices'?

Best,
Sandra

**Comment by finnlindgren** [comment, marked as answer]
You shouldn't need to do much. Below I'll assume your input data `data` is an `sf` object with observations `y` at point locations stored as `sf` geometry. For `model=spde_model`, inlabru would automatically find the mesh and make the required component mapper, but in your case you'll need to specify that manually:
``` r
spde_model <- inla.spde2.matern(spde_mesh, alpha=2)

Q <- inla.spde2.precision(spde_model, theta=Q_params)

comp <-  ~ field(geometry, model = "generic0", Cmatrix = Q, 
   mapper = bru_mapper(spde_mesh),
                group = year_idx, control.group = list(model = "iid")) + ... other covariates ...

fitted_model <- bru(comp, formula = y ~ ., data = data, ... other settings ...)
```
(note: you can also do `comp <- y ~ ...` and leave out the `formula` argument when calling `bru()`; it will be inferred from the component definition)

You can read more about inlabru model and component definitions in https://arxiv.org/abs/2407.00791

**Reply by saerbar** [reply]
That's great, easy to implement and I got it to work, so thank you!
